src/Symfony/Component/Finder/Gitignore.php

```diff
@@ -14,120 +14,78 @@
 /**
  * Gitignore matches against text.
  *
+ * @author Michael Voříšek <vorismi3@fel.cvut.cz>
  * @author Ahmed Abdou <mail@ahmd.io>
  */
 class Gitignore
 {
     /**
      * Returns a regexp which is the equivalent of the gitignore pattern.
      *
-     * @return string The regexp
+     * Format specification: https://git-scm.com/docs/gitignore#_pattern_format
      */
     public static function toRegex(string $gitignoreFileContent): string
     {
-        $gitignoreFileContent = preg_replace('/^[^\\\r\n]*#.*/m', '', $gitignoreFileContent);
-        $gitignoreLines = preg_split('/\r\n|\r|\n/', $gitignoreFileContent);
+        $gitignoreFileContent = preg_replace('~(?<!\\\\)#[^\n\r]*~', '', $gitignoreFileContent);
+        $gitignoreLines = preg_split('~\r\n?|\n~', $gitignoreFileContent);
 
-        $positives = [];
-        $negatives = [];
+        $res = self::lineToRegex('');
         foreach ($gitignoreLines as $i => $line) {
-            $line = trim($line);
-            if ('' === $line) {
-                continue;
+            if (\strlen($line) < 2 || '\\' !== substr($line, -2, 1)) {
+                $line = rtrim($line, ' ');
             }
 
-            if (1 === preg_match('/^!/', $line)) {
-                $positives[$i] = null;
-                $negatives[$i] = self::getRegexFromGitignore(preg_replace('/^!(.*)/', '${1}', $line), true);
-
-                continue;
-            }
-            $negatives[$i] = null;
-            $positives[$i] = self::getRegexFromGitignore($line);
-        }
-
-        $index = 0;
-        $patterns = [];
-        foreach ($positives as $pattern) {
-            if (null === $pattern) {
-                continue;
+            if (0 === strpos($line, '!')) {
```

Comment:
```suggestion
            if ($line[0] === '!') {
```

Reply:
no - I do not use strings as arrays

Reply:
well, looking for the position of `!` in the entire string when only the first char matters is not efficient, and a waste of CPU time.
I could suggest using `if (substr($line, 0, 1) === '!')` to fix this, but then, people will suggest replacing the call to the `substr method` by `[0]` which is more efficient. 
So given we know the string is not empty, please, use `$line[0]`. this is not about dealing with array vs string. But using the most effective way allowed by PHP to check the first char of a string.

Reply:
well, looking for the position of `!` in the entire string when only the first char matters is not efficient, and a waste of CPU time.
I could suggest using `if (substr($line, 0, 1) === '!')` to fix this, but then, people will suggest replacing the call to the `substr method` by `[0]` which is more efficient. 
So given we know the string is not empty, please, use `$line[0]`. this is not about dealing with array vs string. But using the most effective way allowed by PHP to check the first char of a string.

Reply:
It is not empry, `substr($line, 0, 1) === '!'` is good. Code updated.